subprojects/core/src/test/groovy/org/gradle/api/internal/file/archive/TarFileTreeTest.java

```diff
@@ -157,4 +157,16 @@ public void readsTarFileWithNullPermissions() {
 
         assertVisitsPermissions(tree, expected);
     }
+
+    @Test
+    public void doesNotOverwriteFilesOnSecondVisit() {
+        rootDir.file("file1.txt").write("content");
+        rootDir.tarTo(tarFile);
+
+        assertVisits(tree, toList("file1.txt"), new ArrayList<String>());
+        File unpackedFile = expandDir.listFiles()[0].listFiles()[0];
+        unpackedFile.setWritable(false);
+        assertVisits(tree, toList("file1.txt"), new ArrayList<String>());
+        assertFalse(unpackedFile.canWrite());
```

Comment:
But then the timestamps could be accidentally the same, since the test will often complete much faster than the time resolution. How do we usually avoid that?